### PR TITLE
BUG: Fix groupby quantile segfault

### LINKED ISSUE #27470: BUG: Groupby.quantile segfaults for quantile > 1
Adding tests for #27467 surfaced some bugs:

```python
from pandas import *
df = DataFrame(dict(a=[0, 0, 0, 1, 1, 1], b=range(6)))
g = df.groupby(np.repeat([0,1], 3))
# ok
g.quantile(50)
<segfault>
```
quantile accepts floats in `[0,1]` or arrays of floats. It appears not to validate the range though.

### PR DESCRIPTION
**Pull request by TomAugspurger**
Validate that q is between 0 and 1.

Closes #27470

### CODE REVIEW

File: pandas/_libs/groupby.pyx [hunk 1]
```diff
@@ -719,6 +719,11 @@ def group_quantile(ndarray[float64_t] out,
         ndarray[int64_t] counts, non_na_counts, sort_arr
 
     assert values.shape[0] == N
+
+    if q < 0 or q > 1:
```
Comment (jbrockmendel) [inline pandas/_libs/groupby.pyx]: can this be `not 0 <= q <= 1`

Comment (TomAugspurger) [reply pandas/_libs/groupby.pyx]: Seems like it works with parens.

**Comment by jbrockmendel** [comment]
LGTM

**Review by jbrockmendel — APPROVED** [review]

**Review by jreback — APPROVED** [review]
lgtm

**Comment by jbrockmendel** [comment]
Thanks @TomAugspurger

**Comment by WillAyd** [comment]
Going through the 0.25.3 release the test for this showed up in the diff while cherry-picking. On a second review I don't think this was fully back ported (?). I see the whatsnew note but not the test or the changes to groupby.pyx

Unless I am overlooking something should I pick this up for the 0.25.3 release as well?

**Pull request merged**